packages/multi-test/src/custom_handler.rs

```diff
@@ -1,50 +1,72 @@
-use cosmwasm_std::{Addr, Api, Binary, BlockInfo, Empty, Storage};
+use cosmwasm_std::{Addr, Api, Binary, BlockInfo, Storage};
 
 use anyhow::Result as AnyResult;
 use derivative::Derivative;
 use std::cell::{Ref, RefCell};
+use std::marker::PhantomData;
 use std::ops::Deref;
 use std::rc::Rc;
 
 use crate::AppResponse;
 
 /// Custom message handler trait. Implementor of this trait is mocking environment behavior on
 /// given custom message.
-pub trait CustomHandler<ExecC = Empty, QueryC = Empty> {
+pub trait CustomHandler {
+    /// Custom exec message for this handler
+    type ExecC;
```

Comment:
Huh, I have seen such in eg. Iterator, but I am not sure how they are different in practice than generic traits.

This is needed for the Sudoku?

Reply:
So difference is on the one hand subtle, on the other - fundamental.

When you have a generic trait, then it basically means, that this trait can be implemented on type multiple times, for different generics argument. So if `CustomHandler` is generic trait, you can make your very same type to be handler for multiple types. But what it also means is, that when you call a function on this trait, the if the function doesn't have all its generics in signature, type deduction my fail. Consider:

```rust
// Assume MyCustomHandler implements:
// * `CustomHandler<Empty, Empty>`
// * `CustomHandler<CustomMsg, Empty>`
let handler = MyCustomHandler::new();
// Now: which to call? Both `CustomHandler` implementations are valid here
handler.query(&api, &storage, &block, Empty);
```
 And the problem is even worse than hinting for type in normal types - as when for types you basically typehint once and the type is known for whole context, for traits it is not true. Even if here you call `query` from particular implementation, it doesn't mean that in next line, you don't want to call the other `query` - therefore you need to typehint all your calls and it looks like: `CustomHandler::<Empty, _>::query(&handler, &storage, &block, Empty)`.

On the other hand, associated types are induced from trait - so there is always one trait implementation for type, and types are delivered by this implementation. It is more limiting, but also way simpler for typesystem - types don't need to be ever elided. You see I often put type bounds on those types (`ExecT = ExecT`), but what it means is "if you already check that `CustomHandler` is implemented, make sure, that it is the same that other generic type" (which can also help type elision in case `ExecT` cannot be elided from other context, which is basically often case here).

Perfect example of difference between two approaches are `AsRef` and `Deref` traits which both give returns a borrow to underlying data, but they are different - `AsRef` is generic, `Deref` has associated `Deref::Target`. And they differ on semantics - `AsRef` means, that type can be used in place of borrow to other type (which makes sens to have multiple implementations - `String` can be used both as `&String` and `&str`, and there are more complex scenario, when type can have even more `AsRef impls). `Deref` is designing for dereferencing and it means that it always may be dereferenced to single type, and it never makes sense to have multiple implementation of it. This is a reason, why you always can use `*` on smart pointer (which uses deref internally), but to use `.as_ref` you always have to specify type you are wan to convert to (either directly by type hint, or by using in on unequivocal context).

Now an ancient question - when to use generic traits and when to use associated types? I often see rule of thumb: use generics unless they make you troubles, try associated when generics doesn't work. It is based on idea, that generics are somehow easier to think about I think, but I don't like this rule. My rule is: use generics if it makes sense to implement this trait with different types, otherwise use associated types to avoid type elision problems. Either way - in this very case, the generic implementation I took on first take proofed itself being problematic - because if in your whole test you chose to use only queries or execs, then type elision went crazy and spam of type hints were needed.

Reply:
Thanks for the explanation